### ISSUE TITLE
Add eslint:recommended to eslintrc.js

### ISSUE THREAD

**Issue by kendricktan** [UMAprotocol/protocol#1821]
**Why**
`eslint:recommended` enables checks for common problems. This helps us avoid common pitfalls.

**How**
Add `eslint:recommended` to `extends`

**Resolution**
`eslint:recommended` is in `.eslintrc.js`

**Difficulty Score [1-10]**
3, <1 day

**Comment by mrice32** [comment]
Yeah, I think this is a good idea! I quickly tried adding it and it picks up a bunch of accidentally defined globals and unused variables which is great!

I think we'll have to toy with the config a bit to get it to allow some expected globals and have it work for all of the different environments in our code. Might also make sense to have a separate config for the dapp directory if we want to have distinct rules for nodejs vs browser.

**Comment by adrianmcli** [comment]
> Difficulty Score [1-10]
1, <5 minutes

lol, this made me laugh. Wouldn't it also be prudent to budget in time for making sure the repo passes the config?

**Comment by adrianmcli** [comment]
Handy link for specifying globals in your config: https://eslint.org/docs/user-guide/configuring#specifying-globals

> Might also make sense to have a separate config for the dapp directory if we want to have distinct rules for nodejs vs browser.

I think it makes sense to have package-specific `.eslintrc` if each package is sufficiently different enough (like browser vs node for example).

**Comment by kendricktan** [comment]
> I think we'll have to toy with the config a bit to get it to allow some expected globals and have it work for all of the different environments in our code. Might also make sense to have a separate config for the dapp directory if we want to have distinct rules for nodejs vs browser.

Definitely, I can configure it so it has our commonly used globals in it

> lol, this made me laugh. Wouldn't it also be prudent to budget in time for making sure the repo passes the config?

I forgot about that :+1: Editted original post for realistic estimate

> I think it makes sense to have package-specific .eslintrc wherever possible

Agreed, however not all packages within the repo conform to a single standard (e.g. browser vs node)

**Comment by chrismaree** [comment]
ye, this is a great idea. I'm also curious around how we'd want to split this up between the packages like @mrice32 & @adrianmcli pointed out. Could there be one base config at the root of the repo and then package-specific overrides? 

How does this interplay with prettier?

**Comment by kendricktan** [comment]
> Could there be one base config at the root of the repo and then package-specific overrides?

Yeah this is supported by default me thinks.

> How does this interplay with prettier?

Since we're already using prettier as an `eslint` plugin, it shouldn't affect it too much. If there is too much conflict between the two (I highly doubt it) we can always [turn off the conflicting rules](https://github.com/prettier/eslint-config-prettier). @adrianmcli  might have a better idea since he's my go-to for JS advice.

**Comment by adrianmcli** [comment]
## Re: Extending Configs

Yes, you can extend config files, here's the relevant section:
https://eslint.org/docs/user-guide/configuring#using-a-configuration-file

So you'd have an `.eslintrc` at project root, and then in each individual package (if so desired) can have their own `.eslintrc` that extends from the one at project root.

## Re: Prettier

**A historical note:** It used to be that there were plugins that integrated Prettier _into_ your linter so that it would run Prettier as a part of `lint --fix`. However, that is no longer recommended, in fact the docs have been changed to reflect this:
https://prettier.io/docs/en/integrating-with-linters.html#notes

> Use Prettier for code formatting concerns, and linters for code-quality concerns

In short, the current recommended approach is to have ESLint and Prettier to be completely separate. The only thing you have to do is make sure that the [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) config in your `.eslintrc` so that you disable rules that may conflict with Prettier.

**Comment by daywiss** [comment]
I think its a good idea. I've been on some code bases with really intense eslint/tslint rules and it can start to feel kind of overboard. The main rule i can think of which was really annoying was requiring object properties and imports to be alphabetized. Theres a good reason for this, it helps git merge changes without conflicts, but there are times where this behavior does not make sense. Just be sensible about what kind of rules you want and generally I think this will pick up a lot of issues that could lead to subtle problems.

**Comment by adrianmcli** [comment]
> I've been on some code bases with really intense eslint/tslint rules and it can start to feel kind of overboard.

Totally agree. I believe in very minimal configs, _ideally_ we extend from `eslint:recommended` and add no other rules, or only specific rules that we really _need_.
